### PR TITLE
Add the documentation

### PR DESCRIPTION
**Pull request by FredTheDino**
- Add MatchableValue
- Rewrite lib sylt
- Make the macro even more prettier
- Minor fixes
----
Everything above this line is allready precent in "the other PR".
I couldn't be asked to integrate these changes one after another,
and I felt the needed to be done now. So I did them now. The diff might
be hard to look at though. :(

- Show where we need doc-strings
- Add some nice macro that writes to a file. :3
- Add a simple documentation generator
- I documented the lot!

### CODE REVIEW

File: src/lib_sylt.rs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 use crate::*;
 use crate as sylt;
 
+#[sylt_macro::sylt_doc(dbg, "Writes the type and value of any thing you enter", [One(Value(val))] Type::Void)]
 #[sylt_macro::sylt_link(push, "sylt::lib_sylt")]
 pub fn dbg(values: &[Value], _typecheck: bool) -> Result<Value, RuntimeError> {
```
Comment (sornas) [inline src/lib_sylt.rs:6]: The `push` here looks wrong. Maybe we should be able to catch it in the macro?

File: src/lib_sylt.rs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 use crate::*;
 use crate as sylt;
 
+#[sylt_macro::sylt_doc(dbg, "Writes the type and value of any thing you enter", [One(Value(val))] Type::Void)]
```
Comment (sornas) [inline src/lib_sylt.rs:4]:
```suggestion
#[sylt_macro::sylt_doc(dbg, "Writes the type and value of anything you enter", [One(Value(val))] Type::Void)]
```

File: src/lib_sylt.rs [hunk 3]
```diff
@@ -68,41 +71,47 @@ pub fn prepend(values: &[Value], typecheck: bool) -> Result<Value, RuntimeError>
...
+        )),
+    }
+}
 
 sylt_macro::extern_function!(
     "sylt::lib_sylt"
     sin
-    [Value::Float(t)] -> Type::Float => {
-        Ok(Value::Float(t.sin()))
+    "The sinus function you know and love from trigenometry class"
```
Comment (sornas) [inline src/lib_sylt.rs:94]:
```suggestion
    "The sine function you know and love from trigonometry class"
```

File: src/lib_sylt.rs [hunk 4]
```diff
@@ -68,41 +71,47 @@ pub fn prepend(values: &[Value], typecheck: bool) -> Result<Value, RuntimeError>
...
+        Ok(Float(t.sin()))
     },
 );
 
 sylt_macro::extern_function!(
     "sylt::lib_sylt"
     cos
-    [Value::Float(t)] -> Type::Float => {
-        Ok(Value::Float(t.cos()))
+    "The cosinus function you know and love from trigenometry class"
```
Comment (sornas) [inline src/lib_sylt.rs:103]:
```suggestion
    "The cosine function you know and love from trigonometry class"
```

File: src/lib_sylt.rs [hunk 5]
```diff
@@ -68,41 +71,47 @@ pub fn prepend(values: &[Value], typecheck: bool) -> Result<Value, RuntimeError>
...
+        Ok(Float(t.cos()))
     },
 );
 
 sylt_macro::extern_function!(
     "sylt::lib_sylt"
     as_float
-    [Value::Int(t)] -> Type::Float => {
-        Ok(Value::Float(*t as f64))
+    "Converts the to a float"
```
Comment (sornas) [inline src/lib_sylt.rs:112]:
```suggestion
    "Converts the int to a float"
```

File: src/lib_sylt.rs [hunk 6]
```diff
@@ -152,7 +161,7 @@ pub fn pop(values: &[Value], typecheck: bool) -> Result<Value, RuntimeError> {
     }
 }
 
-
+#[sylt_macro::sylt_doc(inf, "Returns an infinet iterator, spitting out the value you give it", [One(Value(val))] Type::Iter)]
```
Comment (sornas) [inline src/lib_sylt.rs:164]:
```suggestion
#[sylt_macro::sylt_doc(inf, "Returns an infinite iterator, spitting out the value you give it", [One(Value(val))] Type::Iter)]
```

File: src/lingon_sylt.rs [hunk 7]
```diff
@@ -141,202 +72,280 @@ macro_rules! game {
 sylt_macro::extern_function!(
     "sylt::lingon_sylt"
     l_update
+    "Updates the engine, needs to be called once per frame"
```
Comment (sornas) [inline src/lingon_sylt.rs:75]:
```suggestion
    "Updates the engine. Needs to be called once per frame"
```

File: src/lingon_sylt.rs [hunk 8]
```diff
@@ -141,202 +72,280 @@ macro_rules! game {
...
         game!().update(0.0);
-        Ok(Value::Nil)
+        Ok(Nil)
     },
 );
 
 sylt_macro::extern_function!(
     "sylt::lingon_sylt"
     l_render
+    "Draws all render calls to the screen, needs to be called once per frame"
```
Comment (sornas) [inline src/lingon_sylt.rs:86]:
```suggestion
    "Draws all render calls to the screen. Needs to be called once per frame"
```

File: src/lingon_sylt.rs [hunk 9]
```diff
@@ -141,202 +72,280 @@ macro_rules! game {
...
+    [Two(Float(sx), Float(sy))] -> Type::Void => {
+        game!().renderer.camera.scale_by(*sx as f32, *sy as f32);
+        Ok(Nil)
     },
 );
 
 sylt_macro::extern_function!(
     "sylt::lingon_sylt"
     l_gfx_particle_new
+    "Creates a new particle system. Note specially the return type, don't edit the return value."
```
Comment (sornas) [inline src/lingon_sylt.rs:275]:
```suggestion
    "Creates a new particle system. Note specially the return type. Don't edit the return value."
```

File: src/lingon_sylt.rs [hunk 10]
```diff
@@ -397,102 +418,114 @@ particle_prop! { l_gfx_particle_end_alpha, end_alpha }
...
         use lingon::input::{Device, Button};
         let button = if let Some(button) = Button::from_string(button) {
             button
         } else {
             return error!("l_bind_button", "'{}' is an invalid button", button);
         };
 
-        game!().input.bind(Device::Button(*controller as u32, button), String::clone(name));
-        Ok(Value::Nil)
+        game!().input.bind(Device::Button(*controller as u32, button), std::string::String::clone(name));
```
Comment (sornas) [inline src/lingon_sylt.rs:502]: Consider `name.clone()` instead, both here and above/below.

**Review by sornas — APPROVED** [review]
What about the failing CI?

File: src/lib_sylt.rs (hunk 1 above)
Comment (FredTheDino) [reply src/lib_sylt.rs:6]: That was what the "no duplicates" were supposed to catch...

File: src/lingon_sylt.rs (hunk 10 above)
Comment (FredTheDino) [reply src/lingon_sylt.rs:502]: I couldn't. And I got frustrated....

**Pull request merged**